File: poh1/poh1_2.scala
//import scala.collection.JavaConversions._
object Main extends App{
	val Array(n,d) = readLine().split(" ").map(_.toInt)
	val _v=new Array[Int](1000001)
	val v=new Array[Int](n)
	for(i<-0 to n-1)_v(readInt())+=1
	var i=0
	for(j<-0 to 1000000)for(k<-0 to _v(j)-1){
		v(i)=j
		i+=1
	}
	for(i<-0 to d-1){
		val m=readInt()
		var idx=n//java.util.Arrays.binarySearch(v,m-v(0)+1)
		if(idx<0)idx=(~idx)
		var r=0
		var j=0
		var k=idx-1
		while(r<m&&j<k&&v(j)+v(j+1)<=m){
				while(v(j)+v(k)>m)k-=1
				if(r<v(j)+v(k))r=v(j)+v(k)
				j+=1
		}
		println(r)
	}
}

File: poh1/poh1_2.fs
open System
let arg:String array=Console.ReadLine().Split(' ')
let n=int(arg.[0])
let d=int(arg.[1])
let _v:int array=Array.zeroCreate(1000001)
let v:int array=Array.zeroCreate(n)
for i in 0..n-1 do
 let x=int(Console.ReadLine())
 _v.[x]<-_v.[x]+1
let mutable x=0
for j in 0..1000000 do
 for k in 0.._v.[j]-1 do
  v.[x]<-j
  x<-x+1
for i in 0..d-1 do
 let m=int(Console.ReadLine())
 let mutable idx=Array.BinarySearch(v,m-v.[0]+1)
 if idx<0 then
  idx<-(~~~idx)
 let mutable r=0
 let mutable j=0
 let mutable k=idx-1
 while r<m && j<k && v.[j]+v.[j+1]<=m do
  while v.[j]+v.[k]>m do
   k<-k-1
  if r<v.[j]+v.[k] then
   r<-v.[j]+v.[k]
  j<-j+1
 Console.WriteLine(r)